Mathlib/Topology/UniformSpace/MultipliableTendstoUniformly.lean

```diff
@@ -0,0 +1,88 @@
+
+/-!
+# Uniform convergence of products of functions
+
+We gather some results about the uniform convergence of infinite products, in particular those of
+the form `∏' i, (1 + f i x)` for a sequence `f` of functions.
+-/
+
+open Filter Function Complex Finset
+
+variable {α β ι : Type*} [UniformSpace β] [AddGroup β] [UniformAddGroup β] [LinearOrder β]
+  [OrderTopology β] [AddLeftMono β] [AddRightMono β]
+
+lemma TendstoUniformlyOn.eventually_forall_lt {f : ι → α → β} {p : Filter ι} {g : α → β}
+    {K : Set α} {u v : β} (huv : u < v) (hf : TendstoUniformlyOn f g p K) (hg : ∀ x ∈ K, g x ≤ u) :
+    ∀ᶠ i in p, ∀ x ∈ K, f i x < v := by
+  simp only [tendstoUniformlyOn_iff_tendsto, uniformity_eq_comap_neg_add_nhds_zero,
+    tendsto_iff_eventually, eventually_comap, Prod.forall] at *
+  conv at hf => enter [2]; rw [eventually_iff_exists_mem]
+  have hf2 := hf (fun x ↦ -x.1 + x.2 < -u + v) ⟨_, (isOpen_gt' (-u + v)).mem_nhds (by simp [huv]),
+    fun y hy a b hab => (hab.symm ▸ hy :)⟩
+  rw [eventually_prod_principal_iff] at *
+  filter_upwards [hf2] with i hi x hx
+  simpa using add_lt_add_of_le_of_lt (hg x hx) (hi x hx)
+
+lemma TendstoUniformlyOn.eventually_forall_le {f : ι → α → β} {p : Filter ι} {g : α → β}
+    {K : Set α} {u v : β} (huv : u < v) (hf : TendstoUniformlyOn f g p K) (hg : ∀ x ∈ K, g x ≤ u) :
+    ∀ᶠ i in p, ∀ x ∈ K, f i x ≤ v := by
+  filter_upwards [hf.eventually_forall_lt huv hg] with i hi x hx using (hi x hx).le
+
+lemma TendstoUniformlyOn.comp_cexp {p : Filter ι} {f : ι → α → ℂ} {g : α → ℂ}
+    {K : Set α} (hf : TendstoUniformlyOn f g p K) (hg : BddAbove <| (fun x ↦ (g x).re) '' K) :
+    TendstoUniformlyOn (cexp ∘ f ·) (cexp ∘ g) p K := by
+  obtain ⟨v, hv⟩ : ∃ v, ∀ x ∈ K, (g x).re ≤ v := hg.imp fun _ h ↦ by simpa [mem_upperBounds] using h
+  have : ∀ᶠ i in p, ∀ x ∈ K, (f i x).re ≤ v + 1 := hf.re.eventually_forall_le (lt_add_one v) hv
+  refine (UniformContinuousOn.cexp _).comp_tendstoUniformlyOn_eventually (by simpa) ?_ hf
+  simpa using fun a ha ↦ (hv a ha).trans (lt_add_one v).le
+
+lemma tendstoUniformlyOn_tprod_of_clog {f : ι → α → ℂ} {K : Set α}
+    (h : ∀ x ∈ K, Summable fun i => log (f i x))
+    (hf : TendstoUniformlyOn (fun s a => ∑ i ∈ s, log (f i a))
+    (fun a => ∑' i, log (f i a)) atTop K) (hfn : ∀ x ∈ K, ∀ i, f i x ≠ 0)
+    (hg : BddAbove <| (fun x => (∑' n : ι, log (f n x)).re) '' K) :
+    TendstoUniformlyOn (fun s a => ∏ i ∈ s, f i a) (fun a => ∏' i, f i a) atTop K := by
+  suffices H : TendstoUniformlyOn (fun s a => ∏ i ∈ s, f i a)
+       (cexp ∘ fun a ↦ ∑' i, log (f i a)) atTop K by
+        apply TendstoUniformlyOn.congr_right H
+        exact fun x hx ↦ (cexp_tsum_eq_tprod (fun n ↦ f n x) (hfn x hx) (h x hx))
+  apply TendstoUniformlyOn.congr (hf.comp_cexp hg)
+  filter_upwards with s i hi using by simp [exp_sum, fun y ↦ exp_log (hfn i hi y)]
+
+/-- This is the version of `tendstoUniformlyOn_tprod_of_clog` for infinite products of with terms
+of the form `1 + f n x`. -/
```

Comment:
This does not seem to be an accurate description of the contents of the lemma. It also changes many things at once relative to the previous lemma (the shape of the terms has changed, the uniform bound `u` has made an appearance, the index set is `ℕ` rather than an arbitrary `ι`, etc); it might be better to try to make these steps one at a time, rather than all at once, in order to offer the maximum flexibility to others who want to use these lemmas in future